blockchain/v2/scheduler.go

```diff
@@ -0,0 +1,715 @@
+		return fmt.Errorf("received block %d from peer %s without being requested", height, peerID)
+	}
+
+	pendingTime, ok := sc.pendingTime[height]
+	if !ok || now.Sub(pendingTime) <= 0 {
+		return fmt.Errorf("clock error: block %d received at %s but requested at %s",
+			height, pendingTime, now)
+	}
+
+	// This may divide by zero - needs to measure every 10s blocks (see v0 and v1),
+	// for now changed to nsecs.
+	peer.lastRate = size / now.Sub(pendingTime).Nanoseconds()
+
+	sc.setStateAtHeight(height, blockStateReceived)
+	delete(sc.pendingBlocks, height)
+	delete(sc.pendingTime, height)
+
+	sc.receivedBlocks[height] = peerID
+
+	return nil
+}
+
+func (sc *scheduler) markPending(peerID p2p.ID, height int64, time time.Time) error {
+	state := sc.getStateAtHeight(height)
+	if state != blockStateNew {
+		return fmt.Errorf("block %d should be in blockStateNew but is %s", height, state)
+	}
+
+	peer, ok := sc.peers[peerID]
+	if !ok {
+		return fmt.Errorf("cannot find peer %s", peerID)
+	}
+
+	if peer.state != peerStateReady {
+		return fmt.Errorf("cannot schedule %d from %s in %s", height, peerID, peer.state)
+	}
+
+	if height > peer.height {
+		return fmt.Errorf("cannot request height %d from peer %s that is at height %d",
+			height, peerID, peer.height)
+	}
+
+	sc.setStateAtHeight(height, blockStatePending)
+	sc.pendingBlocks[height] = peerID
+	// XXX: to make this more accurate we can introduce a message from
+	// the IO routine which indicates the time the request was put on the wire
+	sc.pendingTime[height] = time
+
+	return nil
+}
+
+func (sc *scheduler) markProcessed(height int64) error {
+	state := sc.getStateAtHeight(height)
+	if state != blockStateReceived {
+		return fmt.Errorf("cannot mark height %d received from block state %s", height, state)
+	}
+
+	sc.height++
+	delete(sc.receivedBlocks, height)
+	delete(sc.blockStates, height)
+	sc.addNewBlocks()
+
+	return nil
+}
+
+func (sc *scheduler) allBlocksProcessed() bool {
+	return sc.height >= sc.maxHeight()
+}
+
+// returns max peer height or the last processed block, i.e. sc.height
+func (sc *scheduler) maxHeight() int64 {
+	max := sc.height - 1
+	for _, peer := range sc.peers {
+		if peer.state != peerStateReady {
+			continue
+		}
+		if max < peer.height {
+			max = peer.height
+		}
+	}
+	return max
+}
+
+// lowest block in sc.blockStates with state == blockStateNew or -1 if no new blocks
+func (sc *scheduler) nextHeightToSchedule() int64 {
+	var min int64 = math.MaxInt64
+	for height, state := range sc.blockStates {
+		if state == blockStateNew && height < min {
+			min = height
+		}
+	}
+	if min == math.MaxInt64 {
+		min = -1
+	}
+	return min
+}
+
+func (sc *scheduler) pendingFrom(peerID p2p.ID) []int64 {
+	var heights []int64
+	for height, pendingPeerID := range sc.pendingBlocks {
+		if pendingPeerID == peerID {
+			heights = append(heights, height)
+		}
+	}
+	return heights
+}
+
+func (sc *scheduler) selectPeer(height int64) (p2p.ID, error) {
```

Comment:
does it select the peer with min pending requests?

Reply:
Yes. I think motivation for the generic name Is to extend peer selection to use the download speed or other factors.